### PR TITLE
[RISCV][CostModel] Add getRISCVInstructionCost() to TTI for Cost…

### PR DESCRIPTION
**Pull request by arcbbb**
…Kind

Instruction cost for CodeSize and Latency/RecipThroughput can be very differnet. Considering the diversity of CostKind and vendor-specific cost, and how they are spread across various TTI functions, it's becoming quite a challenge to handle. This patch adds an interface getRISCVInstructionCost to address it.

### CODE REVIEW

**Comment by preames** [comment]
What's the actual problem you're hoping to solve here?  Where are you going with this direction?

**Review by wangpc-pp — COMMENTED** [review]
I think this makes how the cost is calculated much more consistent and clearer.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:38]: What about making it `getRISCVInstructionCost(ArrayRef<RISCVInstruction> Insts, MVT VT, TTI::TargetCostKind CostKind)` and calculate the cost sum of `Insts`?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:71]: Is the plan to use these in a later patch?

**Comment by arcbbb** [comment]
> What's the actual problem you're hoping to solve here? Where are you going with this direction?
@preames 
Two problem I would like to solve in this way:

1. Centralization of instruction cost updates:
  The primary goal is to simplify the process of modifying these costs.
For example, consider the case of instructions like "VSLIDE" and "VMERGE". Introducing vendor-specific cost updates for these instructions currently requires extensive modifications scattered throughout the code
2. Improve the way costs are computed for TCK_CodeSize, TCK_Latency, and TCK_RecipThroughput.
For the VMV_VX instruction, the cost calculations are differentiated based on the cost metric type: 
it's set to 1 for TCK_CodeSize, scales with LMULCost for TCK_Latency, and could potentially be calculated as `DIV(LMULCost, NumUnit)` for TCK_RecipThroughput. 

And in this way I would like to update TTI cost functions of shuffle, cast, reduction, and arithmetic instruction in follow-up patches.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Fixed. Thanks!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.h
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.h]: Sounds good, I have updated based on it.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Updated!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:339]: If there is only one element, we can drop the braces.

Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Fixed. Thanks!

**Comment by arcbbb** [comment]
ping

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:518]: This FIXME is stale.

**Review by wangpc-pp — COMMENTED** [review]
I'm OK with this change, but please wait for others' approval.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Fixed. Thanks!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:41]: I assume we've now changed the behavior of CostKind==TTI::TCK_CodeSize by removing all the getLMULCost calls? Can we test that?

Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:41]: I update the related shuffle tests.

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:41]: Then it is no longer a NFC. :-)

Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:41]: Yeah, I just forgot this.. I'll rename the title.

**Comment by arcbbb** [comment]
ping

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:497]: Is this slidedown.vx or slidedown.vi? It's a shuffle so the mask should be know at compile time right?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:57]: .vx and .vi slides are quite different. The .vi instructions know at decode time how far to slide so hardware can know early which source DLEN pieces are needed for each DLEN piece of the result.

.vx requires the sources to determined after looking at the scalar register.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Yes, the index should be known and I can check if it is a 5-bit imm for .vi.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (arcbbb) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:59]: Yes, I think it would be good to expand getVSlideCost to getVSlideVXCost and getVSlideVICost. Fixed.

**Review by topperc — APPROVED** [review]
LGTM

**Comment by vitalybuka** [comment]
This breaks a few bots with https://lab.llvm.org/buildbot/#/builders/168/builds/17710

Should we revert it?

**Comment by topperc** [comment]
> This breaks a few bots with https://lab.llvm.org/buildbot/#/builders/168/builds/17710
> 
> Should we revert it?

Please revert it

**Comment by arcbbb** [comment]
> This breaks a few bots with https://lab.llvm.org/buildbot/#/builders/168/builds/17710
> 
> Should we revert it?

@vitalybuka Thanks! I am still trying to reproduce this in my local.

**Pull request merged**
